### PR TITLE
891: Scalable mapping

### PR DESCRIPTION
**Pull request by md5**
This PR updates the dashboard map to work better for missions with large numbers of responses.

1. Uses [`MarkerClusterer`](https://github.com/googlemaps/js-marker-clusterer) to group together responses
2. Uses a smaller icon for individual responses
3. Only shows the most recent 1,000 responses

### CODE REVIEW

File: app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    @map = new google.maps.Map($('div.response_locations')[0], {
+      mapTypeId: google.maps.MapTypeId.ROADMAP,
+      zoom: 1,
+      streetViewControl: false,
+      draggableCursor: 'pointer'
+    })
+
+    # create the marker clusterer
+    mc = new MarkerClusterer(@map, [], {
+      imagePath: 'https://google-maps-utility-library-v3.googlecode.com/svn/trunk/markerclusterer/images/m'
```
Comment (md5) [inline app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee]: @hooverlunch Let me know if you'd like me to try to vendor these marker images.

File: app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      split = loc.split(' ')
+      lat = parseFloat(split[0])
+      lng = parseFloat(split[1])
+
+      # create marker
+      p = new google.maps.LatLng(lat, lng)
+      m = new google.maps.Marker({
+        position: p,
+        title: I18n.t('activerecord.models.response.one') + ' #' + response_id,
+        icon: 'https://storage.googleapis.com/support-kms-prod/SNP_2752125_en_v0',
```
Comment (md5) [inline app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee]: Same here. We may want to vendor this icon.

File: app/views/welcome/index.html.erb [hunk 3]
```diff
@@ -62,6 +62,11 @@
     <%# no need to cache this bit either because there's nothing here (done by JS) %>
     <h2 class="interior"><%= t('dashboard.response_locations') %></h2>
     <div class="response_locations"></div>
+    <% if @location_answers_count > @location_answers.size %>
```
Comment (md5) [inline app/views/welcome/index.html.erb]: Hopefully the styling for this is OK. I've got a 13" screen on my MBP, so it's difficult for me to test the styling on the dashboard due to its fancy attempts to recalculate element sizes.

**Comment by md5** [comment]
BTW, @hooverlunch I did figure out how to control the text used for the `MarkerClusterer` markers. While looking through the code yesterday, I noticed you could call `setCalculator` to give it a custom "calculator" function. This allows you to control the index of the marker cluster image (they come in sizes 1-5) and the text that appears. I don't think that `setCalculator` is documented, but there are a few StackOverflow and other mentions of people using it.

File: app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee (hunk 2 above)
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee]: Yes I think yes to both.

File: app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee [hunk 4]
```diff
@@ -0,0 +1,117 @@
+# ELMO.Views.DashboardMapView
+#
+# View model for the dashboard map
+class ELMO.Views.DashboardMapView extends Backbone.View
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/dashboard_map_view.js.coffee]: Yay Backbone!

**Comment by md5** [comment]
This needs some more work. Will reopen this PR or another when it's ready.

File: app/models/answer.rb [hunk 5]
```diff
@@ -34,17 +34,22 @@ class Answer < ActiveRecord::Base
...
+    end
+
+    # return an AR relation
+    where.not(:value => nil)
+      .joins(:questioning => :question)
+      .where(:questions => { :qtype_name => 'location' })
+      .joins(:response)
+      .where(:responses => response_conditions)
+      .order('responses.updated_at DESC')
+      .paginate(:page => 1, :per_page => 1000)
```
Comment (hooverlunch) [inline app/models/answer.rb]: Very nice.

**Comment by hooverlunch** [comment]
This looks fantastic. So clean and nice, a real addition to the codebase.

I do wonder about MarkerClusterer though. I think we'd be better off without it. GMaps should be able to handle rendering the 1000 small dots directly. I think it's more striking visually that way. I just tried it on my own machine and I like it better without. So lets axe it, unless you have a reason for keeping it.

Then please go ahead and merge.

Thanks!

**Comment by md5** [comment]
I agree about `MarkerClusterer`. Besides the fact that 1,000 small dots actually works pretty well, the `MarkerClusterer` itself had some wonkiness. I was seeing cases where it wouldn't zoom correctly on click, as well as cases where it zoomed out instead of in on click. There's also a weird interaction between the bounds setting code and the clusterer that causes the centers of some clusters to end up right on the edge of the view, effectively clipping the cluster marker in half.

I'll remove the `MarkerClusterer` code and merge.

**Pull request merged**